run-benchmark should allow profiles to be taken during browser benchmark runs

xw opened the pull request:
#### 619e29b635b5478321472aa69a23645a8dd018c2
<pre>
run-benchmark should allow profiles to be taken during browser benchmark runs
<a href="https://bugs.webkit.org/show_bug.cgi?id=247496">https://bugs.webkit.org/show_bug.cgi?id=247496</a>
rdar://101969614

Reviewed by Dewei Zhu.

We&apos;d like the ability to take profiles in browser benchmarks. To do this, we&apos;ll need to add a context (defined in the driver) to take profiles during test iterations, and add an argument to the script that toggles this feature on.

In addition:
 - Added support for signposts in OSX driver, and refactored `_launch_url_with_custom_path` to be able to use a previously modified environment.
 - Added an `iteration` argument to `_run_one_test`, as this information may be useful for labelling profiles later on.
 - Clarify and rename references to PGO profiles - previously, we had use the term &quot;profile generation&quot; across run-benchmark to refer to PGO profile generation (this has since been replaced with the phrase &quot;PGO profile generation&quot;).

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
(BenchmarkBuilder.__init__):
(BenchmarkBuilder.__enter__):
* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
(BenchmarkRunner._run_benchmark):
(BenchmarkRunner.execute):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver):
(BrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_safari_driver.py:
(OSXSafariDriver.prepare_env):
(OSXSafariDriver.launch_url):
(OSXSafariDriver._launch_url_with_custom_path):
(OSXSafariDriver.close_browsers):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(parse_args):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webdriver_benchmark_runner.py:
(WebDriverBenchmarkRunner._run_one_test):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner._get_result):
(WebServerBenchmarkRunner):

Canonical link: <a href="https://commits.webkit.org/256544@main">https://commits.webkit.org/256544@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd406d34296e6d67769b584384f2b3b35db4aa7d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5004 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28792 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105326 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165633 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99733 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5083 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33760 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88131 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101162 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101411 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3738 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82362 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30791 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/99336 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85601 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87507 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73620 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39495 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19045 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37188 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20362 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4527 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41253 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43005 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43219 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39613 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
```diff
@@ -51,6 +52,11 @@ def __enter__(self):
                 self._apply_patch(self._plan[patch_file_key])
+            if self._enable_signposts:
+                if 'signpost_patch' in self._plan:
+                    self._apply_patch(self._plan['signpost_patch'])
```
What are we applying a patch to?

xw replied:
We're patching the benchmark itself - see https://github.com/WebKit/WebKit/pull/6232

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
```diff
@@ -22,7 +22,7 @@ def istext(a):
     name = 'benchmark_runner'
 
-    def __init__(self, plan_file, local_copy, count_override, build_dir, output_file, platform, browser, browser_path, scale_unit=True, show_iteration_values=False, device_id=None, diagnose_dir=None, profile_output_directory=None):
+    def __init__(self, plan_file, local_copy, count_override, build_dir, output_file, platform, browser, browser_path, scale_unit=True, show_iteration_values=False, device_id=None, diagnose_dir=None, pgo_profile_output_dir=None, profile_output_dir=None):
```
It's a bit weird to add a new argument _before_ an old one, and what's the motivation for changing `profile_output_directory` to `profile_output_dir`?

xw replied:
Two things are happening here:
1. `profile_output_directory` is being renamed to `pgo_profile_output_dir` and refers to where we put profiles generated by the `--generate-pgo-profile` flag, we want to be more specific now that we have two things in run-benchmark that refer to "profiles".
2. `profile_output_dir` is new and refers to where we put profiles generated by the `--profile` flag.

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_safari_driver.py:
```diff
@@ -20,31 +20,28 @@ def prepare_env(self, config):
                 env[key] = value
-
+        if self._enable_signposts:
+            env['__XPC_WEBKIT_SIGNPOSTS_ENABLED'] = "1"
```
I'm a bit surprised we don't need `WEBKIT_SIGNPOSTS_ENABLED` too, this wouldn't enable signposts in the UI process, I think.

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
```diff
@@ -47,14 +46,29 @@ def _get_result(self, test_url):
+            if enable_profiling:
+                context = self._browser_driver.profile(profile_filename)
+            else:
+                try:
```
`webkitcorepy` has a `NullContext`.

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
```diff
@@ -47,14 +46,29 @@ def _get_result(self, test_url):
+                try:
+                    from contextlib import nullcontext
+                except ImportError:
+                    # Python 2 does not have contextlib.nullcontext, so we must define our own.
```
Where are we running this script as Python 2?

xw replied:
webkitpy EWS has a Python2 component: https://ews-build.webkit.org/#/builders/5/builds/99106/steps/12/logs/stdio

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
```diff
@@ -55,7 +55,8 @@ def config_argument_parser():
     parser.add_argument('--no-adjust-unit', dest='scale_unit', action='store_false', help="Don't convert to scientific notation.")
     parser.add_argument('--show-iteration-values', dest='show_iteration_values', action='store_true', help="Show the measured value for each iteration in addition to averages.")
-    parser.add_argument('--generate-profiles', dest='generate_profiles', action='store_true', help="Collect LLVM profiles for PGO, and copy them to the diagnose directory.")
+    parser.add_argument('--generate-pgo-profiles', '--generate-profiles', dest="generate_pgo_profiles", action='store_true', help="Collect LLVM profiles for PGO, and copy them to the diagnostics directory.")
```
Don't we need to change the references to `--generate-profiles`  on `Tools/Scripts/build-and-collect-pgo-profiles`?

xw replied:
We do, but that's for a separate patch. This is backwards compatible, so `--generate-profiles` will still collect PGO profiles (for now).

dewei-zhu replied:
I think they should be in one patch, otherwise, in the gap between this patch and that patch, invoking this script with `--generate-profiles` will fail.

xw replied:
`--generate-profiles` will still work after this is landed though, I was planning on a separate patch to change `build-and-collect-pgo-profiles` to the new flag and remove the old one after this patch lands. Or unless you are saying it would be a better idea to do it all in one go?

dewei-zhu replied:
Ah, missed `'--generate-profiles'`

JonWBedard commented on Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_safari_driver.py:
```diff
@@ -20,30 +20,27 @@ def prepare_env(self, config):
+            if re.match(r'^__XPC_', key):
                 env[key] = value
+        if self._enable_signposts:
+            env['__XPC_WEBKIT_SIGNPOSTS_ENABLED'] = '1'
```
Just want to make sure you're clear on what `__XPC_` is doing.

Prefixing an environment variable with `__XPC_` will forward that environment variable to all XPC processes. In the case of WebKit, that means we're forwarding `WEBKIT_SIGNPOSTS_ENABLED` to the networking, WebContent and GPU processes. Notably, `WEBKIT_SIGNPOSTS_ENABLED` _will not_ be set in the UI process. That might be desirable, just wanted to call out that fact.

xw replied:
Not forwarding to the UI process should be fine, I believe we mostly care about WebContent in the context of our browser benchmarks.

xw committed:
run-benchmark should allow profiles to be taken during browser benchmark runs
https://bugs.webkit.org/show_bug.cgi?id=247496
rdar://101969614

Reviewed by Dewei Zhu.

We'd like the ability to take profiles in browser benchmarks. To do this, we'll need to add a context (defined in the driver) to take profiles during test iterations, and add an argument to the script that toggles this feature on.

In addition:
 - Added support for signposts in OSX driver, and refactored `_launch_url_with_custom_path` to be able to use a previously modified environment.
 - Added an `iteration` argument to `_run_one_test`, as this information may be useful for labelling profiles later on.
 - Clarify and rename references to PGO profiles - previously, we had use the term "profile generation" across run-benchmark to refer to PGO profile generation (this has since been replaced with the phrase "PGO profile generation").

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
(BenchmarkBuilder.__init__):
(BenchmarkBuilder.__enter__):
* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner.__init__):
(BenchmarkRunner._run_benchmark):
(BenchmarkRunner.execute):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/browser_driver.py:
(BrowserDriver):
(BrowserDriver.profile):
* Tools/Scripts/webkitpy/benchmark_runner/browser_driver/osx_safari_driver.py:
(OSXSafariDriver.prepare_env):
(OSXSafariDriver.launch_url):
(OSXSafariDriver._launch_url_with_custom_path):
(OSXSafariDriver.close_browsers):
* Tools/Scripts/webkitpy/benchmark_runner/run_benchmark.py:
(config_argument_parser):
(parse_args):
(run_benchmark_plan):
* Tools/Scripts/webkitpy/benchmark_runner/webdriver_benchmark_runner.py:
(WebDriverBenchmarkRunner._run_one_test):
* Tools/Scripts/webkitpy/benchmark_runner/webserver_benchmark_runner.py:
(WebServerBenchmarkRunner.__init__):
(WebServerBenchmarkRunner._get_result):
(WebServerBenchmarkRunner):

Canonical link: https://commits.webkit.org/256544@main
